ls1intum/Artemis
# Multiple improvements for programming exercises

### Checklist

### Motivation and Context
`Trigger all builds` and `build and test after due date` did not work for all participations in large courses due to problems with `resume exercise` and with transactions. We carefully inspected the problems and implemented many small improvements so that these features work as expected

### Description
* Remove most `Transactional` annotations for programming exercises. We don't really need transactions here. This makes the execution faster and less error-prone in terms of concurrency issues
* Fix concurrency issues during `Trigger all builds` and `build and test after due date`
* Change the order of empty commits for resume exercise
* Make sure students can always resume programming exercises, even if a step fails
* Improve error handling
* Filter all branches except master and Artemis setup commits to prevent wrong submissions
* Use the student / instructor user name and email for commits in the online code editor and when generating the exercise
* Handle go to build plan on client to prevent null pointer exceptions

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
4. Start exercise in Courses
5. Clean up the build plan
6. Run Trigger All in Participations
7. Check that all builds have been executed

**Note:** All other aspects of programming exercises need to be tested too.
We need to simulate the trigger all button with 1000 test users on the test server
